pkg/bgpv1/manager/manager.go

```diff
@@ -529,6 +523,256 @@ func (m *BGPRouterManager) Stop() {
 		s.Server.Stop()
 	}
 
+	for _, i := range m.BGPInstances {
+		i.Router.Stop()
+	}
+
 	m.Servers = make(LocalASNMap)
+	m.BGPInstances = make(LocalInstanceMap)
 	m.running = false
 }
+
+// ReconcileInstances is a API for configuring the BGP Instances from the
+// desired CiliumBGPNodeConfig resource.
+//
+// ReconcileInstances will evaluate BGP instances to be created, removed and
+// reconciled.
+func (m *BGPRouterManager) ReconcileInstances(ctx context.Context,
+	nodeObj *v2alpha1api.CiliumBGPNodeConfig,
+	ciliumNode *v2api.CiliumNode) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// use a reconcileDiff to compute which BgpServers must be created, removed
+	// and reconciled.
+	rd := newReconcileDiffV2(ciliumNode)
+
+	if nodeObj == nil {
+		m.withdrawAllV2(ctx, rd)
+		return nil
+	}
+
+	l := m.Logger.WithFields(logrus.Fields{
+		types.BGPNodeConfigLogField: nodeObj.Name,
+	})
+
+	err := rd.diff(m.BGPInstances, nodeObj)
+	if err != nil {
+		return err
+	}
+
+	if rd.empty() {
+		l.Debug("BGP instance up-to-date with CiliumBGPNodeConfig")
+		return nil
+	}
+	l.WithField("diff", rd.String()).Debug("Reconciling BGP instances")
+
+	if len(rd.register) > 0 {
+		if err := m.registerV2(ctx, rd); err != nil {
+			return err
+		}
+	}
+	if len(rd.withdraw) > 0 {
+		m.withdrawV2(ctx, rd)
+	}
+	if len(rd.reconcile) > 0 {
+		if err := m.reconcileV2(ctx, rd); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// registerV2 instantiates and configures BGP Instance(s) as instructed by the provided
+// work diff.
+func (m *BGPRouterManager) registerV2(ctx context.Context, rd *reconcileDiffV2) error {
+	var instancesWithError []string
+	for _, name := range rd.register {
+		var config *v2alpha1api.CiliumBGPNodeInstance
+		var ok bool
+		if config, ok = rd.seen[name]; !ok {
+			m.Logger.WithField(types.InstanceLogField, name).Error("Work diff (add) contains unseen instance, skipping")
+			instancesWithError = append(instancesWithError, name)
+			continue
+		}
+		if rErr := m.registerBGPInstance(ctx, config, rd.ciliumNode); rErr != nil {
+			// we'll log the error and attempt to register the next instance.
+			m.Logger.WithField(types.InstanceLogField, name).WithError(rErr).Errorf("Error while registering new BGP instance")
```

Comment:
Here, we log an error with the field logger with the instance name. So, we don't need to include the instance name in the error message returned from functions called from `registerBGPInstance`. I see some errors containing instance names. Same for `reconcileV2`.